subsys/usb/device_next/class/bt_hci.c

```diff
@@ -0,0 +1,491 @@
+#define USB_BT_HCI_PROTOCOL		0x01
+
+#define BT_HCI_DEFAULT_BULK_EP_MPS	0	/* Get maximum supported */
+#define BT_HCI_DEFAULT_INT_EP_MPS	64
+#define BT_HCI_CLASS_ENABLED		0
+#define BT_HCI_ACL_RX_ENGAGED		1
+
+static K_FIFO_DEFINE(bt_hci_rx_queue);
+static K_FIFO_DEFINE(bt_hci_tx_queue);
+
+/*
+ * REVISE: global (bulk, interrupt, iso) specific pools would be more
+ * RAM usage efficient.
+ */
+NET_BUF_POOL_FIXED_DEFINE(bt_hci_ep_pool,
+			  3, 512,
+			  sizeof(struct udc_buf_info), NULL);
+/* HCI RX/TX threads */
+static K_KERNEL_STACK_DEFINE(rx_thread_stack, CONFIG_BT_HCI_TX_STACK_SIZE);
+static struct k_thread rx_thread_data;
+static K_KERNEL_STACK_DEFINE(tx_thread_stack, 512);
+static struct k_thread tx_thread_data;
+
+struct bt_hci_data {
+	struct net_buf *acl_buf;
+	uint16_t acl_len;
+	struct k_sem sync_sem;
+	atomic_t state;
+};
+
+/*
+ * Make supported device request visible for the stack
+ * bRequest 0x00 and bRequest 0xE0.
+ */
+static const struct usbd_cctx_vendor_req bt_hci_vregs =
+	USBD_VENDOR_REQ(0x00, 0xe0);
+
+struct usbd_bt_hci_desc {
+	struct usb_if_descriptor if0;
+	struct usb_ep_descriptor if0_int_ep;
+	struct usb_ep_descriptor if0_out_ep;
+	struct usb_ep_descriptor if0_in_ep;
+} __packed;
+
+static uint8_t bt_hci_get_int_in(struct usbd_class_node *const c_nd)
+{
+	struct usbd_bt_hci_desc *desc = c_nd->data->desc;
+
+	return desc->if0_int_ep.bEndpointAddress;
+}
+
+static uint8_t bt_hci_get_bulk_in(struct usbd_class_node *const c_nd)
+{
+	struct usbd_bt_hci_desc *desc = c_nd->data->desc;
+
+	return desc->if0_in_ep.bEndpointAddress;
+}
+
+static uint8_t bt_hci_get_bulk_out(struct usbd_class_node *const c_nd)
+{
+	struct usbd_bt_hci_desc *desc = c_nd->data->desc;
+
+	return desc->if0_out_ep.bEndpointAddress;
+}
+
+struct net_buf *bt_hci_buf_alloc(const uint8_t ep)
+{
+	struct net_buf *buf = NULL;
+	struct udc_buf_info *bi;
+
+	buf = net_buf_alloc(&bt_hci_ep_pool, K_NO_WAIT);
+	if (!buf) {
+		return NULL;
+	}
+
+	bi = udc_get_buf_info(buf);
+	memset(bi, 0, sizeof(struct udc_buf_info));
+	bi->ep = ep;
+
+	return buf;
+}
+
+static void bt_hci_tx_sync_in(struct usbd_class_node *const c_nd,
+			      struct net_buf *const bt_buf, const uint8_t ep)
+{
+	struct bt_hci_data *hci_data = c_nd->data->priv;
+	struct net_buf *buf;
+
+	buf = bt_hci_buf_alloc(ep);
+	if (buf == NULL) {
+		LOG_ERR("Failed to allocate buffer");
+		return;
+	}
+
+	net_buf_add_mem(buf, bt_buf->data, bt_buf->len);
+	usbd_ep_enqueue(c_nd, buf);
+	k_sem_take(&hci_data->sync_sem, K_FOREVER);
+	net_buf_unref(buf);
+}
+
+static void bt_hci_tx_thread(void *p1, void *p2, void *p3)
+{
+	struct usbd_class_node *const c_nd = p1;
+
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+
+	while (true) {
+		struct net_buf *bt_buf;
+		uint8_t ep;
+
+		bt_buf = net_buf_get(&bt_hci_tx_queue, K_FOREVER);
+
+		switch (bt_buf_get_type(bt_buf)) {
+		case BT_BUF_EVT:
+			ep = bt_hci_get_int_in(c_nd);
+			break;
+		case BT_BUF_ACL_IN:
+			ep = bt_hci_get_bulk_in(c_nd);
+			break;
```

Comment:
What about ISO in?

Reply:
It is optional and should not be in the first alternate configuration (here: the only one).

Reply:
I'm not sure I understand your comment. ACL is also optional. 

The code even previously mentions ISO, but does not implement anything related to it.

Reply:
> I'm not sure I understand your comment. ACL is also optional.

How is it optional? Please direct me to the relevant Bluetooth specification page/chapter.

BLUETOOTH CORE SPECIFICATION Version 5.3 | Vol 4, Part B page 1667 does not make the ACL Data endpoints optional at all.

After a second read through the Part B: USB Transport Layer it seems that there are two interfaces here:
- First one (the one implemented here) that contains bulk and interrupt endpoints
- Second one that provides scalable isochronous bandwidth

This PR entirely misses the second interface with alternate configurations.

BLUETOOTH CORE SPECIFICATION Version 5.3 | Vol 4, Part B , 2.1.1 Controller descriptors
> The Controller configuration consists of two interfaces. The first interface has
no alternate settings and contains the bulk and interrupt endpoints. The second
interface provides scalable isochronous bandwidth. The recommended
configuration for the second interface has four alternate settings that provide
different bandwidth. The default interface is empty so that the device is capable
of scaling down to zero isochronous bandwidth.

Reply:
> > I'm not sure I understand your comment. ACL is also optional.
> 
> How is it optional? Please direct me to the relevant Bluetooth specification page/chapter.
> 
> BLUETOOTH CORE SPECIFICATION Version 5.3 | Vol 4, Part B page 1667 does not make the ACL Data endpoints optional at all.

That table (2.1) is a recommendation - Not a specification. If you have a advertising (GAP Broadcaster) or scanning (GAP Observer) device only, you don't have any ACL. 

> After a second read through the Part B: USB Transport Layer it seems that there are two interfaces here:
> 
>     * First one (the one implemented here) that contains bulk and interrupt endpoints
> 
>     * Second one that provides scalable isochronous bandwidth
> 
> 
> This PR entirely misses the second interface with alternate configurations.

Will it be added as part of this PR then, or done later?

Reply:
> That table (2.1) is a recommendation - Not a specification. If you have a advertising (GAP Broadcaster) or scanning (GAP Observer) device only, you don't have any ACL.

While there won't be any ACL data in this case, the HCI USB Transport layer (which is implemented here, except for the isochronous interface) must have endpoints for ACL. The text I quoted previously contains
> The first interface has no alternate settings and contains the bulk and interrupt endpoints.

The interrupt endpoint is used for HCI events and bulk endpoints are used for ACL Data (as noted in 2.3 BULK ENDPOINTS EXPECTATIONS). There is no possibility for "interface with only interrupt endpoint".

Reply:
> What about ISO in?

What you probably mean will go in the future through bulk endpoint.
This is porting of `subsys/usb/device/class/bluetooth.c ` without h4 bloat (will be added later), nothing more.